echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/pagerduty/PagerDutyNotificationService.groovy

```diff
@@ -50,17 +58,91 @@ class PagerDutyNotificationService implements NotificationService {
 
   @Override
   void handle(Notification notification) {
+    Map details = new LinkedHashMap()
+    details.put("details", notification.additionalContext.details)
+
+    Map<String, String> errors = new HashMap()
+    Map<String, String> pdErrors = new HashMap()
+
     notification.to.each {
-      pagerDuty.createEvent(
-        "Token token=${token}",
-        new PagerDutyService.PagerDutyCreateEvent(
-          service_key: it,
-          client: "Spinnaker (${notification.source.user})",
-          description: notification.additionalContext.message
+      try {
+        Map response = pagerDuty.createEvent(
+          "Token token=${token}",
+          new PagerDutyService.PagerDutyCreateEvent(
+            service_key: it,
+            client: "Spinnaker (${notification.source.user})",
+            description: notification.additionalContext.message,
+            details: details
+          )
         )
-      )
 
-      log.info('Sent page {} {}', kv('serviceKey', it), kv('message', notification.additionalContext.message))
+        if (response.status.equals("success")) {
+          // Page successful
+          log.info('Sent page {} {}', kv('serviceKey', it), kv('message', notification.additionalContext.message))
+        } else {
+          pdErrors.put(it, (String)response.message)
+        }
+      } catch (RetrofitError error) {
+        String errorMessage = error.response.reason
+        TypedInput responseBody = error.response.getBody()
+        if (responseBody != null) {
+          PagerDutyErrorResponseBody errorResponse = mapper.readValue(new String(((TypedByteArray)responseBody).getBytes()), PagerDutyErrorResponseBody)
+          if (errorResponse.errors && errorResponse.errors.size() > 0) {
+            errorMessage = errorResponse.errors.join(", ")
+          }
+        }
+        log.error('Failed to send page {} {} {}',  kv('serviceKey', it), kv('message', notification.additionalContext.message), kv('error', errorMessage))
+        errors.put(it, errorMessage)
+      }
+    }
+
+    // If some errors occurred
+    if (errors.size() > 0 || pdErrors.size() > 0) {
+      String message = ""
+      String pagerDutyError = pdErrors.collect { "${it.key}: ${it.value}" }.join(", ")
+      if (pagerDutyError != null && !pagerDutyError.isEmpty()) {
+        message += " ${pagerDutyError}."
+      }
+      String requestError = errors.collect { "${it.key}: ${it.value}" }.join(", ")
+      if (requestError != null && !requestError.isEmpty()) {
+        message += " ${requestError}."
+      }
+      // If some successes occurred
+      if (errors.size() + pdErrors.size() < notification.to.size()) {
+        throw new PartialPagerDutySuccessException(message)
```

Comment:
I suspect that normally we would actually just throw one type of exception and pass in the full message.

It's a tad strange to pass in `message` but have it actually only form part of the exception message.

Reply:
Yeah, good call. Was separate for different properties, but rolled it all into message.